[GStreamer] Reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE

philn opened the pull request:
#### 8cff56dfd0887ba4bb96093f621714ceeaa3ecc6
<pre>
[GStreamer] Reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE
<a href="https://bugs.webkit.org/show_bug.cgi?id=302526">https://bugs.webkit.org/show_bug.cgi?id=302526</a>

Reviewed by Adrian Perez de Castro and Xabier Rodriguez-Calvar.

Remove unsafe-buffer-usage pragma usage in more places by using span and avoiding memcpy(),
memset() and strcmp() calls.

Also, driving-by, fixing a couple issues in the SDP time repeat attribute handling of the
SDPStringBuilder.

* Source/WebCore/Modules/mediastream/gstreamer/GStreamerDataChannelHandler.cpp:
(WebCore::GStreamerDataChannelHandler::onMessageString):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp:
(WebCore::GStreamerMediaEndpoint::doSetRemoteDescription):
(WebCore::GStreamerMediaEndpoint::processSDPMessage):
(WebCore::GStreamerMediaEndpoint::updatePtDemuxSrcPadCaps):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerWebRTCUtils.cpp:
(WebCore::generateCertificate):
(WebCore::setSsrcAudioLevelVadOn):
(WebCore::SDPStringBuilder::SDPStringBuilder):
(WebCore::extractMidAndRidFromRTPBuffer):
* Source/WebCore/platform/audio/gstreamer/FFTFrameGStreamer.cpp:
(WebCore::FFTFrame::FFTFrame):
* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
(WebCore::registerWebKitGStreamerElements):
(WebCore::GstMappedFrame::componentData const):
(WebCore::GstMappedFrame::planeData const):
* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.h:
* Source/WebCore/platform/graphics/gstreamer/ImageGStreamerSkia.cpp:
(WebCore::ImageGStreamer::ImageGStreamer):
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::notifyPlayerOfTrack):
(WebCore::MediaPlayerPrivateGStreamer::handleMessage):
(WebCore::MediaPlayerPrivateGStreamer::setVisibleInViewport):
(WebCore::MediaPlayerPrivateGStreamer::createVideoSinkGL):
* Source/WebCore/platform/graphics/gstreamer/VideoFrameGStreamer.cpp:
(WebCore::copyPlane):
(WebCore::VideoFrame::copyTo):
* Source/WebCore/platform/graphics/gstreamer/WebKitAudioSinkGStreamer.cpp:
(webKitAudioSinkConfigure):
* Source/WebCore/platform/gstreamer/GStreamerCodecUtilities.cpp:
(WebCore::GStreamerCodecUtilities::parseHEVCProfile):
* Source/WebCore/platform/mediastream/libwebrtc/gstreamer/GStreamerVideoFrameLibWebRTC.cpp:
(WebCore::GStreamerVideoFrameLibWebRTC::ToI420):

Canonical link: <a href="https://commits.webkit.org/303317@main">https://commits.webkit.org/303317@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/648b4bc5c8d534ba6da9b46c1b4109b55b7849d7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/132033 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/4525 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/43050 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/139547 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/83941 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/133903 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/4468 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/4287 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/100932 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/83941 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/134979 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/3179 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/118251 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/81722 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/3076 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/952 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/82767 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/111831 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/36373 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/142194 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/4195 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/36952 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/109301 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/4276 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/161/builds/3655 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/109474 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27722 "Built successfully and passed tests") | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/164/builds/3181 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/114528 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/57416 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/4249 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/32923 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/4080 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/67695 "Built successfully") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/4340 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/4208 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aperezdc commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp:
```diff
@@ -902,7 +900,7 @@ void GStreamerMediaEndpoint::doSetRemoteDescription(const RTCSessionDescription&
         processSDPMessage(&message, [this](unsigned, StringView mid, const auto* media) {
             const char* mediaType = gst_sdp_media_get_media(media);
-            m_mediaForMid.set(mid.toString(), g_str_equal(mediaType, "audio") ? RealtimeMediaSource::Type::Audio : RealtimeMediaSource::Type::Video);
+            m_mediaForMid.set(mid.toString(), equalIgnoringASCIICase(mediaType, "audio"_s) ? RealtimeMediaSource::Type::Audio : RealtimeMediaSource::Type::Video);
```
Is it really okay to change those comparisons from case-sensitive to case-insensitive?

philn replied:
Yes it's OK.

calvaris replied:
Turn mediaType into a CStringView and you'll be able to use the == operator on the _s .

aperezdc commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -1214,11 +1212,12 @@ void MediaPlayerPrivateGStreamer::notifyPlayerOfTrack()
-    g_object_get(m_pipeline.get(), numberOfTracksProperty.toCString().data(), &numberOfTracks, nullptr);
+    StringBuilder builder;
+    builder.append("n-"_s, typeName);
+    auto numberOfTracksProperty = builder.toString();
```
Any reason to not use `makeString()`? I think the following would work:


```suggestion
    auto numberOfTracksProperty = makeString("n-"_s, typeName);
```

aperezdc commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -1231,13 +1230,14 @@ void MediaPlayerPrivateGStreamer::notifyPlayerOfTrack()
-    getPadProperty.printf("get-%s-pad", typeName);
+    StringBuilder getPadPropertyBuilder;
+    getPadPropertyBuilder.append("get-"_s, typeName, "-pad"_s);
+    auto getPadProperty = getPadPropertyBuilder.toString();
```
Same here, likely `makeString()` would do:


```suggestion
    auto getPadProperty = makeString("get-"_s, typeName, "-pad"_s);
```

aperezdc reviewed (commented):
LGTM overall, but I suppose it wouldn't hurt if someone else with their multimedia hat on takes also a look. I left a comment/suggestion.

aperezdc reviewed (approved):
Thanks for the update, this is nicer now.

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp:
```diff
@@ -1073,7 +1071,7 @@ void GStreamerMediaEndpoint::processSDPMessage(const GstSDPMessage* message, Fun
         const auto* media = gst_sdp_message_get_media(message, mediaIndex);
         const char* mediaType = gst_sdp_media_get_media(media);
-        if (!g_str_equal(mediaType, "audio") && !g_str_equal(mediaType, "video"))
+        if (!equalIgnoringASCIICase(mediaType, "audio"_s) && !equalIgnoringASCIICase(mediaType, "video"_s))
```
Ditto

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp:
```diff
@@ -2705,7 +2703,7 @@ void GStreamerMediaEndpoint::updatePtDemuxSrcPadCaps(GstElement* ptDemux, GstPad
     for (int j = 0; j < gst_structure_n_fields(s2); j++) {
         const char* name = gst_structure_nth_field_name(s2, j);
-        if (!g_str_equal(name, "media") && !g_str_equal(name, "payload") && !g_str_equal(name, "clock-rate") && !g_str_equal(name, "encoding-name"))
+        if (!equalIgnoringASCIICase(name, "media"_s) && !equalIgnoringASCIICase(name, "payload"_s) && !equalIgnoringASCIICase(name, "clock-rate"_s) && !equalIgnoringASCIICase(name, "encoding-name"_s))
```
ditto

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerWebRTCUtils.cpp:
```diff
@@ -744,7 +745,7 @@ void setSsrcAudioLevelVadOn(GstStructure* structure)
 
         const char* uri = g_value_get_string(value);
-        if (!g_str_equal(uri, GST_RTP_HDREXT_BASE "ssrc-audio-level"))
+        if (!equalIgnoringASCIICase(uri, GST_RTP_HDREXT_BASE "ssrc-audio-level"))
```
ditto

philn replied:
I'll keep this change because CStringView doesn't have a comparison operator for string literals.

calvaris replied:
It does for ASCIILiterals, which is what you need.

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerWebRTCUtils.cpp:
```diff
@@ -1087,42 +1090,44 @@ GRefPtr<GstCaps> extractMidAndRidFromRTPBuffer(const GstMappedRtpBuffer& buffer,
-            if (rid)
-                gst_caps_set_simple(mediaCaps.get(), "a-rid", G_TYPE_STRING, rid.get(), nullptr);
+        if (!mid.empty()) {
+            CString midString(mid);
```
ditto

philn replied:
I'll keep this change because CStringView doesn't have a constructor for spans.

calvaris replied:
It will, soon.

Anyway, decide by yourself if a memory copy is worth saving a couple of size calculations. IMHO, it is not.

You decide neither option is good, I'd not transform the span into a CString either, I would just keep the span and maybe cast if needed when setting.

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerWebRTCUtils.cpp:
```diff
@@ -1087,42 +1090,44 @@ GRefPtr<GstCaps> extractMidAndRidFromRTPBuffer(const GstMappedRtpBuffer& buffer,
+            gst_caps_set_simple(mediaCaps.get(), "a-mid", G_TYPE_STRING, midString.data(), nullptr);
 
+            if (!rid.empty()) {
+                CString ridString(rid);
```
ditto

philn replied:
ditto!

calvaris replied:
re-re-ditto! :laughing:

calvaris commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -2191,8 +2187,8 @@ void MediaPlayerPrivateGStreamer::handleMessage(GstMessage* message)
-                    const gchar* klassStr = gst_element_get_metadata(element, "klass");
-                    if (strstr(klassStr, "Sink") && strstr(klassStr, "Video")) {
+                    auto klassStr = CStringView::unsafeFromUTF8(gst_element_get_metadata(element, "klass"));
+                    if (containsIgnoringASCIICase(klassStr.span(), "Sink"_s)  && containsIgnoringASCIICase(klassStr.span(), "Video"_s)) {
```
You don't need to make this ascii case unaware, right?

calvaris commented on Source/WebCore/platform/graphics/gstreamer/WebKitAudioSinkGStreamer.cpp:
```diff
@@ -56,57 +56,55 @@ WEBKIT_DEFINE_TYPE_WITH_CODE(WebKitAudioSink, webkit_audio_sink, GST_TYPE_BIN,
-            return GST_PAD_PROBE_OK;
-        }), nullptr, nullptr);
-        return true;
+    auto value = StringView::fromLatin1(g_getenv("WEBKIT_GST_ENABLE_AUDIO_MIXER"));
```
give this var a more meaningful name

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp:
```diff
@@ -2704,10 +2702,11 @@ void GStreamerMediaEndpoint::updatePtDemuxSrcPadCaps(GstElement* ptDemux, GstPad
             continue;
-        gst_structure_set_value(s, name, gst_structure_get_value(s2, name));
+        const char* nameString = name.utf8();
+        gst_structure_set_value(s, nameString, gst_structure_get_value(s2, nameString));
```
`name.utf8()` should work in the second line as well! Why are you doing this? If it's about performance, you don't need to worry. It should be inlined.

calvaris commented on Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
```diff
@@ -570,17 +570,15 @@ void registerWebKitGStreamerElements()
         // unless the WEBKIT_GST_ENABLE_LEGACY_VAAPI environment variable is set to 1.
-        const char* enableLegacyVAAPIPlugin = getenv("WEBKIT_GST_ENABLE_LEGACY_VAAPI");
-        if (!enableLegacyVAAPIPlugin || !strcmp(enableLegacyVAAPIPlugin, "0")) {
+        auto enableLegacyVAAPIPlugin = StringView::fromLatin1(g_getenv("WEBKIT_GST_ENABLE_LEGACY_VAAPI"));
```
For new code, even when using StringView is perfectly ok, I think CStringView is a lighter class. No strong opinion on this though.

calvaris commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -3997,8 +3993,8 @@ void MediaPlayerPrivateGStreamer::setVisibleInViewport(bool isVisible)
     // this using an environment variable, set from the webkitpy glib port sub-classes.
-    const char* allowPlaybackOfInvisibleVideos = g_getenv("WEBKIT_GST_ALLOW_PLAYBACK_OF_INVISIBLE_VIDEOS");
-    if (!isVisible && allowPlaybackOfInvisibleVideos && !strcmp(allowPlaybackOfInvisibleVideos, "1"))
+    auto allowPlaybackOfInvisibleVideos = StringView::fromLatin1(g_getenv("WEBKIT_GST_ALLOW_PLAYBACK_OF_INVISIBLE_VIDEOS"));
```
ditto!

calvaris commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -4149,8 +4145,8 @@ MediaPlayer::MovieLoadType MediaPlayerPrivateGStreamer::movieLoadType() const
 {
-    const char* disableGLSink = g_getenv("WEBKIT_GST_DISABLE_GL_SINK");
-    if (disableGLSink && !strcmp(disableGLSink, "1")) {
+    auto disableGLSink = StringView::fromLatin1(g_getenv("WEBKIT_GST_DISABLE_GL_SINK"));
```
ditto

philn committed:
[GStreamer] Reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE
https://bugs.webkit.org/show_bug.cgi?id=302526

Reviewed by Adrian Perez de Castro and Xabier Rodriguez-Calvar.

Remove unsafe-buffer-usage pragma usage in more places by using span and avoiding memcpy(),
memset() and strcmp() calls.

Also, driving-by, fixing a couple issues in the SDP time repeat attribute handling of the
SDPStringBuilder.

* Source/WebCore/Modules/mediastream/gstreamer/GStreamerDataChannelHandler.cpp:
(WebCore::GStreamerDataChannelHandler::onMessageString):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp:
(WebCore::GStreamerMediaEndpoint::doSetRemoteDescription):
(WebCore::GStreamerMediaEndpoint::processSDPMessage):
(WebCore::GStreamerMediaEndpoint::updatePtDemuxSrcPadCaps):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerWebRTCUtils.cpp:
(WebCore::generateCertificate):
(WebCore::setSsrcAudioLevelVadOn):
(WebCore::SDPStringBuilder::SDPStringBuilder):
(WebCore::extractMidAndRidFromRTPBuffer):
* Source/WebCore/platform/audio/gstreamer/FFTFrameGStreamer.cpp:
(WebCore::FFTFrame::FFTFrame):
* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
(WebCore::registerWebKitGStreamerElements):
(WebCore::GstMappedFrame::componentData const):
(WebCore::GstMappedFrame::planeData const):
* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.h:
* Source/WebCore/platform/graphics/gstreamer/ImageGStreamerSkia.cpp:
(WebCore::ImageGStreamer::ImageGStreamer):
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::notifyPlayerOfTrack):
(WebCore::MediaPlayerPrivateGStreamer::handleMessage):
(WebCore::MediaPlayerPrivateGStreamer::setVisibleInViewport):
(WebCore::MediaPlayerPrivateGStreamer::createVideoSinkGL):
* Source/WebCore/platform/graphics/gstreamer/VideoFrameGStreamer.cpp:
(WebCore::copyPlane):
(WebCore::VideoFrame::copyTo):
* Source/WebCore/platform/graphics/gstreamer/WebKitAudioSinkGStreamer.cpp:
(webKitAudioSinkConfigure):
* Source/WebCore/platform/gstreamer/GStreamerCodecUtilities.cpp:
(WebCore::GStreamerCodecUtilities::parseHEVCProfile):
* Source/WebCore/platform/mediastream/libwebrtc/gstreamer/GStreamerVideoFrameLibWebRTC.cpp:
(WebCore::GStreamerVideoFrameLibWebRTC::ToI420):

Canonical link: https://commits.webkit.org/303317@main
